pkg/bootstrap/platform/azure.go

```diff
@@ -0,0 +1,170 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package platform
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	AzureMetadataEndpoint = "http://169.254.169.254"
+	InstanceURL           = AzureMetadataEndpoint + "/metadata/instance"
+	DefaultAPIVersion     = "2019-08-15"
+	SysVendorPath         = "/sys/class/dmi/id/sys_vendor"
+	MicrosoftIdentifier   = "Microsoft Corporation"
+)
+
+var (
+	APIVersion        = DefaultAPIVersion
+	updateVersionOnce = sync.Once{}
+	getAPIVersions    = func() string {
+		return metadataRequest("")
+	}
+	getAzureMetadata = func() string {
+		return metadataRequest(fmt.Sprintf("api-version=%s", APIVersion))
+	}
+
+	azureTagsFn = func(e *azureEnv) map[string]string {
```

Comment:
instead of having these fns take pointers to `azureEnv`, can we just make them fns with the azureEnv receiver?

```golang
func (a azureEnv) Tags() map[string]string { ...
func (a azureEnv) Location() string { ...
func (a azureEnv) Zone() string {...
```

Since they all seem to operate on `computeMetadata`, we can still swap that value out in tests without loss of function, and this would clean things up a bit.

For instance, `Locality()` becomes:

```golang
func (a *azureEnv) Locality() *core.Locality {
  return &core.Locality{
     Region: a.Location(),
     Zone: a.Zone(),
  }
}
```

Reply:
Is there a way to do that while keeping them easily mockable for the test functions? Functions defined as variables have everything passed through with parameters, so I think the above would be difficult to test.

Reply:
Instead of mocks, we could just set `computeMetadata` (and/or `networkMetadata`) to have whatever we want right, and validate the expected behavior, right?

They don't have to be exposed methods, but I feel like they should have the `azureEnv` receiver.

Reply:
Fixed.